include/seqan3/core/copyable_wrapper.hpp

```diff
@@ -0,0 +1,246 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2021, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2021, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::copyable_wrapper.
+ * \author Enrico Seiler <enrico.seiler AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+#include <optional>
+
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+/*!\brief Helper concept for seqan3::copyable_wrapper.
+ * \ingroup core
+ * \see https://en.cppreference.com/w/cpp/ranges/copyable_wrapper
+ * \noapi{Exposition only.}
+ */
+template <typename t>
+concept boxable = std::copy_constructible<t> && std::is_object_v<t>;
+
+/*!\brief Utility wrapper that behaves like std::optional but makes the type conform with the std::copyable concept.
+ * \ingroup core
+ * \see https://en.cppreference.com/w/cpp/ranges/copyable_wrapper
+ */
+template <typename t>
+class copyable_wrapper : public std::optional<t>
+{
+public:
+    using std::optional<t>::optional; //!< Use std::optional constructors.
+    using std::optional<t>::operator=; //!< Use std::optional assignment operators.
+    constexpr copyable_wrapper(copyable_wrapper const &) = default; //!< Defaulted.
+    constexpr copyable_wrapper(copyable_wrapper &&) = default; //!< Defaulted.
+    constexpr ~copyable_wrapper() = default; //!< Defaulted.
+
+    //!\brief copyable_wrapper is default constructible, iff the wrapped type is default initialisable.
```

Comment:
Actually, we can default construct with the wrapped type is default constructible. Otherwise, the default ctor of std::optional is used (the optional has no value).